### PR TITLE
[extension/storage/redis_storage] Add a new Redis extension to store data in transit

### LINKED ISSUE #31682: Support external queue system for exporter via extensions
### Component(s)

_No response_

### Is your feature request related to a problem? Please describe.

No

### Describe the solution you'd like

I want the ability to utilize Redis as the exporter queue to support upstream outages for longer period. The current two options of exporter do not fulfill the requirements or have issues with longer outages.

This usecase came up because I needed near 0% data loss on logs over long periods of time because the upstream is known to be down for many mins. The in-memory queue was ruled out because it would lose data on pod restarts. The persistent queue option was promising but it started refusing new data if the queue is full and based on the code it is difficult to disable the refusal which causes loss of new data once the queue is full.

So since I had access to a Redis cluster, I decided to modify the exporter to support Redis as the queue. I was able to test this and get stable data flow, however I was not able to do proper and long testing yet.

Can this feature be implemented via extensions so in future other queue systems could be added as well? I believe the existing queue systems will need to moved to extensions in that case. 

### Describe alternatives you've considered

Alternatives considered
- In-memory Queue: As mentioned, data is lost if the pod restarts due to crashes.
- Persistent Queue: Starts refusing new data when the queue is full.
- Export to Pulsar, receive from Pulsar: This does work, but the 2nd collector still needs one of the above queue during exporting, ending up in same situation.

### Additional context

Config example
```yaml
exporters:
  logging:
  otlp:
    endpoint: "localhost:4217"
    retry_on_failure:
      enabled: true
      initial_interval: 5s
      max_interval: 5s
      max_elapsed_time: 20s
    sending_queue:
      enabled: true
      requeue_enabled: true
      queue_backend: redis
      num_consumers: 10
      queue_size: 10000
      redis:
        address: "localhost:6379"
        backlog_check_interval: 30
        process_key_expiration: 120
        scan_key_size: 10
    tls:
      insecure: true
```
While queue_size is mentioned 10000, it is actually unlimited because redis can scale up based on usage.

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
Add a new extension allowing to store data in transit in a Redis cluster.

**Link to tracking Issue:**
Fixes #31682

**Testing:**
Unit tests

**Documentation:**
README

### CODE REVIEW

**Comment by atoulme** [comment]
Not stale, waiting for review.

File: extension/storage/redisstorage/factory_test.go
Comment (evan-bradley) [inline extension/storage/redisstorage/factory_test.go:23]: Will these be used in the future?

File: extension/storage/redisstorage/factory.go
Comment (evan-bradley) [inline extension/storage/redisstorage/factory.go:17]:
```suggestion
```
We should probably just remove this comment, since I don't think it explains anything that can't be inferred from the function name (and I don't think it's a HostObserver extension anyway).

File: extension/storage/redisstorage/factory.go
Comment (evan-bradley) [inline extension/storage/redisstorage/factory.go:5]:
```suggestion
```
We already have this in `doc.go`, and I think that's a better "entrypoint" file anyway.

File: extension/storage/redisstorage/testdata/config.yaml
Comment (evan-bradley) [inline extension/storage/redisstorage/testdata/config.yaml:7]:
```suggestion
  expiration: 3h

```
Nit

File: extension/storage/redisstorage/extension.go
Comment (evan-bradley) [inline extension/storage/redisstorage/extension.go:58]:
```suggestion
}

var _ storage.Client = redisClient{}
```

**Review by evan-bradley — APPROVED** [review]
I have limited experience with storage extensions, but this looks good to me for a first pass. All of my suggestions are non-blocking.

File: extension/storage/redisstorage/factory_test.go
Comment (atoulme) [reply extension/storage/redisstorage/factory_test.go]: removed

**Comment by atoulme** [comment]
@open-telemetry/collector-contrib-approvers please review

File: extension/storage/redisstorage/config.go
Comment (crobert-1) [inline extension/storage/redisstorage/config.go:13]: Do we want to define a `Validate` function here, or will the default be effective enough?

File: extension/storage/redisstorage/README.md
Comment (crobert-1) [inline extension/storage/redisstorage/README.md:27]: Maybe it's clear to those with Redis experience, but in terms of explaining configuration, `db: 0` is not self-explanatory.

File: extension/storage/redisstorage/extension.go
Comment (crobert-1) [inline extension/storage/redisstorage/extension.go:97]: This should actually call `Close()` on `rc.client`, right?

Comment (atoulme) [reply extension/storage/redisstorage/extension.go:97]: We reuse the redis client from the extension into each redisClient struct, and manage the lifecycle of the connection to the cluster at the level of the extension start/shutdown methods. That's the same pattern we use for dbstorage.

File: extension/storage/redisstorage/config.go
Comment (atoulme) [reply extension/storage/redisstorage/config.go:13]: The default config is enough, we can't validate much.

File: extension/storage/redisstorage/README.md
Comment (atoulme) [reply extension/storage/redisstorage/README.md:24]: Done, added more info.

File: extension/storage/redisstorage/README.md
Comment (atoulme) [reply extension/storage/redisstorage/README.md:32]: Added more info in the README

File: extension/storage/redisstorage/README.md
Comment (crobert-1) [inline extension/storage/redisstorage/README.md:19]:
```suggestion
- `endpoint` (required): The endpoint of the redis instance to connect to. Default: `localhost:6379`
```
Nit for consistency

**Review by crobert-1 — APPROVED** [review]

File: extension/storage/redisstorage/README.md
Comment (andrzej-stencel) [inline extension/storage/redisstorage/README.md:47]: No such thing as `nop` processor as far as I'm aware.

File: extension/storage/redisstorage/config.go
Comment (andrzej-stencel) [inline extension/storage/redisstorage/config.go:4]: The existing storage extensions (dbstorage, filestorage) are one of the few extensions whose package name does not end with `extension` (https://github.com/open-telemetry/opentelemetry-collector-releases/blob/v0.104.0/distributions/otelcol-contrib/manifest.yaml#L36-L37). Should we continue this pattern? Isn't this a mistake that should be fixed in those extensions and not replicated in a new component?

```suggestion
package redisstorageextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/redisstorageextension"
```

File: extension/storage/redisstorage/extension.go
Comment (andrzej-stencel) [inline extension/storage/redisstorage/extension.go:63]: This means that if several collectors (e.g. daemonset pods) use the same Redis instance, they will step on each other's feet, right? Maybe we can use instance ID or something else in a future PR to prevent this?

File: extension/storage/redisstorage/README.md
Comment (atoulme) [inline extension/storage/redisstorage/README.md:47]:
```suggestion
```

**Comment by evan-bradley** [comment]
@atoulme Anything left to do here, or are we good to merge?

**Comment by atoulme** [comment]
Ready to merge

**Pull request merged**
